File: cpp/src/tt.cpp
#include "tt.hpp"

#include <algorithm>
#include <cstddef>

TranspositionTable::TranspositionTable(int size_mb) {
    std::size_t bytes = static_cast<std::size_t>(size_mb) * 1024 * 1024;
    std::size_t want = bytes / sizeof(TTEntry);
    if (want < 1) want = 1;
    // Round down to a power of two so we can index with a bitmask.
    std::size_t pow2 = 1;
    while (pow2 * 2 <= want) pow2 *= 2;
    table_.assign(pow2, TTEntry{});
    mask_ = pow2 - 1;
}

const TTEntry *TranspositionTable::probe(U64 key) const {
    const TTEntry &e = table_[key & mask_];
    return (e.valid && e.key == key) ? &e : nullptr;
}

void TranspositionTable::store(U64 key, int depth, TTFlag flag, double score, Move best_move) {
    TTEntry &e = table_[key & mask_];
    // Depth-preferred replacement: overwrite empty slots, the same position, or
    // a shallower stored search.
    if (!e.valid || e.key == key || depth >= e.depth) {
        e.key = key;
        e.depth = depth;
        e.flag = flag;
        e.score = score;
        e.best_move = best_move;
        e.valid = true;
    }
}

void TranspositionTable::clear() {
    std::fill(table_.begin(), table_.end(), TTEntry{});
}

File: rust/src/tt.rs
//! Transposition table keyed by Zobrist hash. Port of
//! python/engine/transposition.py and cpp/src/tt: a fixed-size power-of-two
//! array (index = key & mask) with depth-preferred replacement.

use crate::types::{Move, NULL_MOVE, U64};

// TTFlag comes from the shared schema; re-export so `eschess::tt::TTFlag` holds.
pub use crate::generated::TTFlag;

#[derive(Clone, Copy)]
pub struct TTEntry {
    pub key: U64,
    pub depth: i32,
    pub flag: TTFlag,
    pub score: f64,
    pub best_move: Move,
    pub valid: bool,
}

impl TTEntry {
    const EMPTY: TTEntry = TTEntry {
        key: 0,
        depth: 0,
        flag: TTFlag::Exact,
        score: 0.0,
        best_move: NULL_MOVE,
        valid: false,
    };
}

pub struct TranspositionTable {
    table: Vec<TTEntry>,
    mask: usize,
}

impl TranspositionTable {
    pub fn new(size_mb: usize) -> TranspositionTable {
        let bytes = size_mb * 1024 * 1024;
        let want = (bytes / std::mem::size_of::<TTEntry>()).max(1);
        let mut pow2 = 1usize;
        while pow2 * 2 <= want {
            pow2 *= 2;
        }
        TranspositionTable {
            table: vec![TTEntry::EMPTY; pow2],
            mask: pow2 - 1,
        }
    }

    #[inline]
    pub fn probe(&self, key: U64) -> Option<&TTEntry> {
        let e = &self.table[(key as usize) & self.mask];
        if e.valid && e.key == key {
            Some(e)
        } else {
            None
        }
    }

    pub fn store(&mut self, key: U64, depth: i32, flag: TTFlag, score: f64, best_move: Move) {
        let e = &mut self.table[(key as usize) & self.mask];
        // Depth-preferred: overwrite empty slots, the same position, or shallower.
        if !e.valid || e.key == key || depth >= e.depth {
            *e = TTEntry {
                key,
                depth,
                flag,
                score,
                best_move,
                valid: true,
            };
        }
    }
}
